File: webapp/src/Client/ChapterRole.ts
import { useEffect, useState } from "react";
import { useCRUD, UseCRUDOptionalProps } from "../CRUD/CRUD";
import { ChapterRoleType } from "./ChapterRoleType";
import { Individual } from "./Individual";

export type Chapter = {
  id: string;
  name: string;
};

export type MutateChapter = {
  name: string;
};

export type ChapterRole = {
  id: string;
  individual_id: string;
  individual: Individual;
  chapter_id: string;
  chapter: Chapter;
  chapter_role_type_id: string;
  chapter_role_type: ChapterRoleType;
  start_date: string;
  end_date?: string;
};

export type MutateChapterRole = {
  practice_id: string;
  chapter_id: string;
  individual_id: string;
  chapter_role_type_id: string;
};

export interface UseChapterRoleProps extends UseCRUDOptionalProps {
  id?: string;
}

export function useChapterRole(props: UseChapterRoleProps) {
  const {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  } = useCRUD<ChapterRole, MutateChapterRole>({
    path: `/api/capability/chapterroles`,
    ...props,
  });

  return {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  };
}

File: cs/capability/src/Entities/ChapterRole.cs
namespace Capability.Entities;

using Base.Entities;

public class ChapterRole : EntityBase
{
  // ChapterRole descibes an individual taking on
  // a particular ChapterRoleType within a 
  // given Chapter
  // 
  // I.e. "John Doe is a Chapter Lead in the 
  // React Front-End Chapter"

  public required int IndividualId { get; set; }

  public required int ChapterRoleTypeId { get; set; }
  public ChapterRoleType ChapterRoleType { get; set; } = null!;

  public required int ChapterId { get; set; }
  public Chapter Chapter { get; set; } = null!;

  public DateTimeOffset StartDate { get; set; } = DateTimeOffset.UtcNow;
  public DateTimeOffset? EndDate { get; set; }
}